### ISSUE TITLE
Rename issparse?

### ISSUE THREAD

**Issue by garrison** [JuliaLang/julia#24645]
... aka Bikeshed and/or improve documentation for `issparse`.  The docstring for `issparse` says

> `issparse(S)`
>
> Returns `true` if `S` is sparse, and `false` otherwise.

but this is ambiguous.  What, precisely, does it mean for `S` (presumably an array) to be sparse?  Here are two (very different) possible definitions:

1. The array has a density of nonzero elements less than some parameter `x` (`x` could have some default value, but this would be rather arbitrary).
2. The array is represented by a some "sparse" data type.

It is the second option that is implemented by `issparse`.  In a way this makes sense, as the function does not take such a parameter `x` (representing the threshold density of nonzero elements).  But at a minimum this should be more clearly explained in the documentation.  (I'm also unable to explain why `issparse(::Diagonal)` and related types return `false` -- perhaps it's a bug, or perhaps I do not understand fully what this function is meant to do.)

On the other hand, the current behavior makes `issparse` different from many other functions I can think of, such as `istril`, `istriu`, `issymmetric`, `ishermitian`, and `isreal`.  Each of these considers the *value* itself rather than the datatype; for instance, `isreal(0im) == true`.

Perhaps a better name for this function, then, would be `issparsetype`?  I can think of downsides to this name (you do not pass a type to it).  Another idea would be `issparsestorage`, but I don't know if I like this either.

So I ask:

1. Do people agree that `issparse` is a poor name?  If it is, can we come up with anything better?
2. What should `issparse(::Diagonal)`, `issparse(::Tridiagonal)`, etc. return?
3. Is this function even useful?  I just searched discourse, and it's never been mentioned there.  In Base,
 it is used only in the concatenation functions in `sparsevector.jl`.  I have not grepped the package ecosystem, but I wonder how frequently it is used there.

**Comment by andreasnoack** [comment]
The methods are very old and most likely inherited from Matlab. In Julia, I think it makes more sense to check if the type is a `SparseMatrixCSC` or `AbstractSparseMatrix` but there is the complication that a this would be false for e.g. `LowerTriangular{SparseMatrixCSC}` for which `issparse` is true.

**Comment by oxinabox** [comment]
`issparse` should be a property of the type only.
Like a trait.

as @andreasnoack  said it can't be just replaced with:

`x isa AbstractSparseMatrix` as it also depends on type parameters for e.g. `LowerTriangular`.


Its currently defined in https://github.com/JuliaLang/julia/blob/ea34968d5fcf5f083cbd6e15a326a8a72e76dfe1/base/sparse/abstractsparse.jl#L13-L23

which is full of:

```
issparse(A::AbstractArray) = false
issparse(S::AbstractSparseArray) = true
issparse(S::Symmetric{<:Any,<:AbstractSparseMatrix}) = true
```

It would be marginally cleaner if it were:

```
issparse(x:T) where {T} = issparse(T)

issparse(::Type{AbstractArray}) = false
issparse(::Type{AbstractSparseArray}) = true
issparse(::Type{Symmetric{<:Any, T}}) where {T} = issparse(T)
```

But is it an useful trait is a good question.

Having it as a trait might be marginally cleaner in https://github.com/JuliaLang/julia/blob/82ae10df362f699f4ded02b8673bdc4315cf1768/base/sparse/sparsevector.jl#L1015

**Comment by garrison** [comment]
Here is a [list of uses of `issparse`](https://gist.github.com/garrison/002ac101d275b0c4ac431ef71d647f6c) in the julia package ecosystem, created using [cloneallpackages](https://github.com/JuliaPackaging/cloneallpackages) and grep.

**Comment by ViralBShah** [comment]
I like the `isa` solution offered above, and if people are generally onboard, it shouldn't be difficult to get this changed in all the packages.

Perhaps FemtoCleaner can automate it all.

**Comment by Matthew-D-Dwyer** [comment]
HI, 

It looks like this was merged with #34308 in Dec 18. Does that mean this issue can be closed? 

Matt.
